extensions/vscode/src/state.ts

```diff
@@ -105,23 +106,88 @@ export class PublisherState implements Disposable {
     );
   }
 
-  getSelectedContentRecord() {
+  async getSelectedContentRecord() {
     const selection = this.getSelection();
     if (!selection) {
       return undefined;
     }
-    return this.findContentRecordByPath(selection.deploymentPath);
+    // try finding it first within our cache
+    const cr = this.findContentRecordByPath(selection.deploymentPath);
+    if (cr) {
+      return cr;
+    }
+    // if not found, then retrieve it and add it to our cache.
+    try {
+      const api = await useApi();
+
+      const response = await api.contentRecords.get(
+        selection.deploymentName,
+        selection.projectDir,
+      );
+      const cr = response.data;
+      if (!isContentRecordError(cr)) {
+        // its not foolproof, but it may help
+        if (!this.findContentRecord(cr.saveName, cr.projectDir)) {
+          this.contentRecords.push(cr);
+        }
+        return cr;
+      }
+      return undefined;
+    } catch (error: unknown) {
+      const code = getStatusFromError(error);
+      if (code !== 404) {
+        // 400 is expected when doesn't exist on disk
+        const summary = getSummaryStringFromError(
+          "getSelectedContentRecord, contentRecords.get",
+          error,
+        );
+        window.showInformationMessage(
+          `Unable to retrieve deployment record: ${summary}`,
+        );
+      }
+      return undefined;
+    }
   }
 
-  getSelectedConfiguration() {
-    const contentRecord = this.getSelectedContentRecord();
+  async getSelectedConfiguration() {
+    const contentRecord = await this.getSelectedContentRecord();
     if (!contentRecord) {
       return undefined;
     }
-    return this.findValidConfig(
+    const c = this.findValidConfig(
       contentRecord.configurationName,
       contentRecord.projectDir,
     );
+    if (c) {
+      return c;
+    }
+    // if not found, then retrieve it and add it to our cache.
+    try {
+      const api = await useApi();
+      const response = await api.configurations.get(
+        contentRecord.configurationName,
+        contentRecord.projectDir,
+      );
+      const c = response.data;
+      // its not foolproof, but it may help
+      if (!this.findConfig(c.configurationName, c.projectDir)) {
+        this.configurations.push(c);
+      }
+      return c;
+    } catch (error: unknown) {
+      const code = getStatusFromError(error);
+      if (code !== 400) {
```

Comment:
Should this be 404? I think the get_configuration API only returns a 400 if the requested directory is outside the workspace.

Reply:
Yes, good catch!  Commit: https://github.com/posit-dev/publisher/pull/2101/commits/417a42a70bd9b88d5b5666a1864b87249a60a859